File: src/core/countersignature.ts
// Copyright 2026 Haridarman Kumaresan
// SPDX-License-Identifier: Apache-2.0
// Licensed under the Apache License, Version 2.0; see LICENSE at repo root.

import { createHash } from "node:crypto";
import { canonicalize } from "./canonical.js";
import { publicKeyFromSecret, signContext, toB64url, utf8, verifyContext } from "./keys.js";
import { isWebAuthnCredential, verifyWebAuthnAssertion, type WebAuthnPolicy } from "./webauthn.js";
import {
  COUNTERSIGN_VERSION,
  COUNTERSIGNATURE_CONTEXT,
  type Countersignature,
  type Decision,
  type Intent,
  type Policy,
} from "./types.js";

/**
 * Canonical form of an `actor` for distinct-approver counting. Raw-string
 * equality lets one human masquerade as several ("alice", "alice ", "Alice",
 * or NFC/NFD twins) and fill a multi-person quorum alone. We fold Unicode (NFC),
 * trim, and casefold: for every shipped channel the address part is numeric
 * (Discord/Telegram-id/WhatsApp) or case-assigned-unique (Slack IDs, email), so
 * casefolding never merges two genuinely distinct approvers — it only defeats
 * case-variant spoofing. Adapters must still key on STABLE ids (never a mutable
 * username) so the identity itself can't be swapped between approvals.
 */
export function normalizeActor(actor: string): string {
  return String(actor).normalize("NFC").trim().toLowerCase();
}

export interface VerifyOptions {
  /**
   * The authority key(s) the verifier trusts for this Route. When provided,
   * verification additionally requires that the receipt was signed by one of
   * them — closing the gap where a self-signed receipt is "valid" but carries
   * no authority. STRONGLY recommended for any party acting on a receipt.
   */
  trustedKeys?: readonly string[] | string;
  /**
   * Deployment policy for verifying passkey (WebAuthn) receipts. REQUIRED to
   * verify a receipt whose `public_key` is a passkey descriptor — without it a
   * passkey receipt cannot be verified and fails closed.
   */
  webauthn?: WebAuthnPolicy;
}

/**
 * Produce a signed Countersignature over an Intent. `authoritySecret` is
 * the base64url ed25519 seed of whoever vouches for the decision — an
 * adapter acting on a human's button press, or the runtime firing a Default.
 */
export function signDecision(
  intent: Intent,
  decision: Decision,
  actor: string,
  authoritySecret: string,
  policy: Policy = "approver",
  /**
   * ISO 8601 time the decision was made. Defaults to now. The timeout Default
   * overrides this with the deadline itself, so its receipt is never stamped
   * before the window it represents even if the wall clock has drifted back.
   */
  timestamp: string = new Date().toISOString(),
): Countersignature {
  const unsigned = {
    countersign: COUNTERSIGN_VERSION,
    intent_id: intent.intent_id,
    decision,
    actor,
    policy,
    timestamp,
    public_key: publicKeyFromSecret(authoritySecret),
  };
  const signature = signContext(authoritySecret, COUNTERSIGNATURE_CONTEXT, canonicalize(unsigned));
  return { ...unsigned, signature };
}

/**
 * Verify a Countersignature.
 *
 * With no options this checks INTEGRITY only: that the embedded `public_key`
 * signed the canonical receipt. Integrity alone is NOT authority — anyone can
 * mint an integrity-valid receipt with their own key. To act on a receipt,
 * pass `trustedKeys` so verification also proves the receipt was signed by an
 * authority you trust for this Route.
 */
export function verifyCountersignature(cs: Countersignature, opts: VerifyOptions = {}): boolean {
  try {
    const { signature, webauthn, ...unsigned } = cs;
    if (typeof signature !== "string" || typeof cs.public_key !== "string") return false;
    if (opts.trustedKeys !== undefined) {
      const trusted = typeof opts.trustedKeys === "string" ? [opts.trustedKeys] : opts.trustedKeys;
      if (!trusted.includes(cs.public_key)) return false;
    }
    const passkey = isWebAuthnCredential(cs.public_key);
    if (passkey || webauthn !== undefined) {
      // Passkey (WebAuthn) receipt: the signature is an authenticator assertion,
      // not a plain ed25519 signature over the receipt. A webauthn block with a
      // non-passkey key (or vice versa) is malformed → reject. Without an RP
      // policy a passkey receipt cannot be verified → fail closed.
      if (!passkey || !webauthn || typeof webauthn !== "object" || !opts.webauthn) return false;
      // The assertion binds to THIS decision via challenge = digest of the
      // canonical receipt (minus signature and webauthn), domain-separated.
      const challenge = toB64url(
        createHash("sha256").update(utf8(`${COUNTERSIGNATURE_CONTEXT}\n${canonicalize(unsigned)}`)).digest(),
      );
      return verifyWebAuthnAssertion(webauthn, signature, {
        credential: cs.public_key,
        expectedChallenge: challenge,
        rpId: opts.webauthn.rpId,
        allowedOrigins: opts.webauthn.allowedOrigins,
        requireUserVerification: opts.webauthn.requireUserVerification,
      });
    }
    return verifyContext(cs.public_key, COUNTERSIGNATURE_CONTEXT, canonicalize(unsigned), signature);
  } catch {
    return false;
  }
}

File: clients/python/countersign/countersignature.py
# Copyright 2026 Haridarman Kumaresan
# SPDX-License-Identifier: Apache-2.0
"""
Countersignature (receipt) build + verify — matches src/core/countersignature.ts.

Raw-ed25519 keyed and vouched receipts are fully supported. Passkey / WebAuthn
receipts are NOT verified by this core client: they fail closed (return False),
exactly as the reference verifier does without an RP policy. WebAuthn support is a
planned follow-up; the raw-ed25519 keyed path here is the base signer it layers on.
"""
import unicodedata

from .canonical import canonicalize
from .constants import COUNTERSIGN_VERSION, COUNTERSIGNATURE_CONTEXT
from .keys import is_canonical_public_key, public_key_from_secret, sign_context, verify_context

_WEBAUTHN_PREFIXES = ("webauthn-ed25519:", "webauthn-p256:")


def normalize_actor(actor: str) -> str:
    """
    Canonical form of an ``actor`` for distinct-approver counting: NFC-fold, trim,
    casefold. Defeats case/format-variant spoofing without merging genuinely distinct
    approvers (every shipped channel's address is numeric or case-assigned-unique).
    """
    return unicodedata.normalize("NFC", str(actor)).strip().lower()


def is_webauthn_credential(key) -> bool:
    return isinstance(key, str) and key.startswith(_WEBAUTHN_PREFIXES)


def is_valid_credential_descriptor(key) -> bool:
    """
    True iff ``key`` is a well-formed passkey credential descriptor. An ed25519
    passkey wraps a canonical raw key (``webauthn-ed25519:<canonical-key>``); a P-256
    passkey carries a non-empty COSE key blob. (Full WebAuthn assertion verification
    is a follow-up; this only validates the descriptor shape used by keyed approvers.)
    """
    if not is_webauthn_credential(key):
        return False
    material = key.split(":", 1)[1]
    if key.startswith("webauthn-ed25519:"):
        return is_canonical_public_key(material)
    return isinstance(material, str) and len(material) > 0


def credential_key_material(key: str) -> str:
    """
    The underlying key material of a credential: a raw ed25519 key is its own
    material; a ``webauthn-<alg>:<key>`` descriptor's material is the embedded key.
    Lets a raw key and its passkey wrapper be recognized as the same key.
    """
    if is_webauthn_credential(key):
        return key.split(":", 1)[1]
    return key


def sign_decision(intent, decision, actor, authority_secret, policy="approver", timestamp=None):
    """Produce a signed Countersignature over an Intent (vouched or keyed)."""
    if timestamp is None:
        from datetime import datetime, timezone

        timestamp = datetime.now(timezone.utc).strftime("%Y-%m-%dT%H:%M:%S.") + (
            f"{datetime.now(timezone.utc).microsecond // 1000:03d}Z"
        )
    unsigned = {
        "countersign": COUNTERSIGN_VERSION,
        "intent_id": intent["intent_id"],
        "decision": decision,
        "actor": actor,
        "policy": policy,
        "timestamp": timestamp,
        "public_key": public_key_from_secret(authority_secret),
    }
    signature = sign_context(authority_secret, COUNTERSIGNATURE_CONTEXT, canonicalize(unsigned))
    return {**unsigned, "signature": signature}


def verify_countersignature(cs, trusted_keys=None, webauthn=None) -> bool:
    """
    Verify a Countersignature.

    With no ``trusted_keys`` this checks INTEGRITY only (the embedded public_key
    signed the canonical receipt) — which is not authority. Pass ``trusted_keys`` so
    verification also proves the signer is an authority you trust for this Route.
    Passkey receipts fail closed here (see module docstring).
    """
    try:
        if not isinstance(cs, dict):
            return False
        signature = cs.get("signature")
        unsigned = {k: v for k, v in cs.items() if k not in ("signature", "webauthn")}
        public_key = unsigned.get("public_key")
        if not isinstance(signature, str) or not isinstance(public_key, str):
            return False
        if trusted_keys is not None:
            trusted = [trusted_keys] if isinstance(trusted_keys, str) else list(trusted_keys)
            if public_key not in trusted:
                return False
        # Passkey (WebAuthn) receipt or a webauthn block: not supported by the core
        # client — fail closed (the reference verifier also fails closed without a policy).
        if is_webauthn_credential(public_key) or cs.get("webauthn") is not None:
            return False
        return verify_context(public_key, COUNTERSIGNATURE_CONTEXT, canonicalize(unsigned), signature)
    except Exception:
        return False
